### ISSUE TITLE
[MNG-7170] Allow to associate pomFile/${basedir} with DefaultProjectBuilder.build(ModelSource, ...)

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#7959]
**[Mickael Istria](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mickael.istria)** opened **[MNG-7170](https://issues.apache.org/jira/browse/MNG-7170?redirect=false)** and commented

In LemMinX-Maven Language Server for pom.xml files, we're building models from "working copy" (unsaved) pom files. We do so by invoking

```
DefaultProjectBuilder.build( ModelSource modelSource, ProjectBuildingRequest request )
```

One pitfall is that this cascades to pomFile being null for the Model, and as a result ```
${basedir}

```is not resolved.
One could expect that the ```
ModelSource2.getLocationURI()
``` could be used to derive the pomFile when possible. LemMinX-Maven happens to properly set this.
```

---

**Affects:** 3.8.1

**Issue Links:**
- [MNG-5669](https://issues.apache.org/jira/browse/MNG-5669) same pom.xml is read multiple times
 (_**"is fixed by"**_)
- [MNG-7215](https://issues.apache.org/jira/browse/MNG-7215) [Regression] Maven Site Plugin cannot resolve parent site descriptor without locale
 (_**"causes"**_)
- [MNG-7216](https://issues.apache.org/jira/browse/MNG-7216) Revert MNG-7170
 (_**"causes"**_)
- [MNG-7218](https://issues.apache.org/jira/browse/MNG-7218) [Regression] o.a.m.model.Build.getSourceDirectory() incorrectly returns absolute dir on 3.8.2
 (_**"causes"**_)

**Remote Links:**
- [GitHub Pull Request #478
](https://github.com/apache/maven/pull/478)
- [Pull Request
](https://github.com/apache/maven/pull/478)

**Comment by jira-importer** [comment]
**[Mickael Istria](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mickael.istria)** commented

Patch https://github.com/apache/maven/commit/5cdb8332f99a36e5a1da202da43e3c7dfbb49322 for MNG-5669 has fixed that for master/Maven 4.

I tried to apply it to maven-3.8.1 but it seems to have important dependencies on some major Maven 4 commits; so it seems tricky to backport it fully. So I've just reincorporate the related changes, added a test and submitted a PR for the maven-3.8.x branch: https://github.com/apache/maven/pull/478

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

By the way, I don't expect that we will spend time on backporting to Maven 3. Focus is fully on Maven 4.

**Comment by jira-importer** [comment]
**[Mickael Istria](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mickael.istria)** commented

> The model itself shouldn't need this information. The Reader of the Model already says `read ModelSource X` so it knows the location.

I agree it would be best. A setter here is probably the cause why the getter returns null in some cases: the issue with getters is that it's never clear who's supposed to call them.

So do I understand it right that there will be no Maven 3.8.2? When is Maven 4.0 expected to be released?

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

I wouldn't gamble on any Maven 3 releases. Maven 3.8.1 was already kind of unintended, but required as an answer to CVE-2021-26291.
I think we're already hoping for about half a year to push the first alpha out, we should be close.

**Comment by jira-importer** [comment]
**[Mickael Istria](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mickael.istria)** commented

I've implemented a workaround: https://github.com/eclipse/lemminx-maven/commit/2e553da35bcc2495887f9d72209c84085ab433eb

Feel free to close as won't fix if 3.8.2 will never happen

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [5a8997312680a4b9c69a129801524691bc546c08](https://gitbox.apache.org/repos/asf?p=maven.git;a=commit;h=5a8997312680a4b9c69a129801524691bc546c08) for `maven-3.8.x` branch.

You can remove your workaround as soon as 3.8.2 is out.

**Comment by jira-importer** [comment]
**[Mickael Istria](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mickael.istria)** commented

Thanks a lot for the review and merge [Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o) !
